Print time for each step of installation

rna-afk opened the pull request:
Adds a new feature which prints a summary of the time it took in
each stage of the installer after the stage has been completed.

Summary is currently supported for the create, wait-for and the
delete command. The summary is printed at the completion of the
command and will show the time elapsed for each of the stages
of the execution.

Create command has the following stages:-
. Infrastructure
. API
. Bootstrap Complete
. Destroy bootstrap complete.
. Cluster Operator Complete.

Destroy command has the following stages:- 
. Cluster
. Bootstrap

The wait-for command has the following stages:-
- Bootstrap-complete command
	. Bootstrap
	. API
- Install-complete command
	. Cluster
	. Console

rna-afk commented:
/approver @abhinavdahiya

rna-afk commented:
/assign @abhinavdahiya

patrickdillon commented:
Can you paste what the new output looks like for both the durations and the timestamps?

> Also adds a new feature that prints the time for each log if the
log level is set to DEBUG setting which will help the user to
see the time information as the installer is being called.

This should probably be in its own commit. Also, I don't think this is worthwhile as users can get the timestamps by running `tail -f ,openshift_install.log` during the install.

abhinavdahiya commented on cmd/openshift-install/main.go:
```diff
@@ -93,7 +93,8 @@ func runRootCmd(cmd *cobra.Command, args []string) {
 		ForceColors:            terminal.IsTerminal(int(os.Stderr.Fd())),
-		DisableTimestamp:       true,
+		FullTimestamp:          level == logrus.DebugLevel,
+		DisableTimestamp:       level != logrus.DebugLevel,
```
I don't think we should be printing the times for debug output to std out.

abhinavdahiya commented on cmd/openshift-install/create.go:
```diff
@@ -87,6 +88,8 @@ var (
 			// FIXME: add longer descriptions for our commands with examples for better UX.
 			// Long:  "",
 			PostRun: func(_ *cobra.Command, _ []string) {
+				totalTimer := metrics.Timer{}
```
What's with this metrics timer? 

Why is the stdlib time not sufficient here.

abhinavdahiya commented:
What's the goal of the change?

If it's tracking stages like infrastructure creation , bootstrap wait, bootstrap destroy, cluster operator install complete. I don't see how this addresses that completely?

The timer is conflating all assets together.

We are sending another metrics timer and it still requires manual start stop elapsed. What is that package providing value?

rna-afk commented:
> Can you paste what the new output looks like for both the durations and the timestamps?
```
Here's a snippet of the output:
DEBUG[2020-03-09T15:02:19-04:00]       Fetching Install Config...             
DEBUG[2020-03-09T15:02:19-04:00]       Reusing previously-fetched Install Config 
DEBUG[2020-03-09T15:02:19-04:00]     Generating Kubeconfig Admin Internal Client... 
DEBUG[2020-03-09T15:02:19-04:00]     Fetching Kubeconfig Kubelet...           
DEBUG[2020-03-09T15:02:19-04:00]       Fetching Certificate (kube-apiserver-complete-server-ca-bundle)... 
DEBUG[2020-03-09T15:02:19-04:00]       Reusing previously-fetched Certificate (kube-apiserver-complete-server-ca-bundle) 
DEBUG[2020-03-09T15:02:19-04:00]       Fetching Certificate (kubelet-client)... 
DEBUG[2020-03-09T15:02:19-04:00]         Fetching Certificate (kubelet-bootstrap-kubeconfig-signer)... 
DEBUG[2020-03-09T15:02:19-04:00]         Generating Certificate (kubelet-bootstrap-kubeconfig-signer)... 
DEBUG[2020-03-09T15:02:19-04:00]       Generating Certificate (kubelet-client)... 
```

patrickdillon commented:
> Here's a snippet of the output:

Can you give output of the duration changes?

rna-afk commented:
> Can you give output of the duration changes?

DEBUG[2020-03-09T15:41:22-04:00]       Generating Baremetal Config CR...      
DEBUG[2020-03-09T15:41:22-04:00]       Fetching Image...                      
DEBUG[2020-03-09T15:41:22-04:00]       Reusing previously-fetched Image       
DEBUG[2020-03-09T15:41:22-04:00]     Generating Openshift Manifests...        
DEBUG[2020-03-09T15:41:22-04:00]     Fetching Proxy Config...                 
DEBUG[2020-03-09T15:41:22-04:00]     Reusing previously-fetched Proxy Config  
DEBUG[2020-03-09T15:41:22-04:00]     Fetching Certificate (admin-kubeconfig-ca-bundle)... 
DEBUG[2020-03-09T15:41:22-04:00]       Fetching Certificate (admin-kubeconfig-signer)... 
DEBUG[2020-03-09T15:41:22-04:00]       Reusing previously-fetched Certificate (admin-kubeconfig-signer) 
DEBUG[2020-03-09T15:41:22-04:00]     Generating Certificate (admin-kubeconfig-ca-bundle)... 
DEBUG[2020-03-09T15:41:22-04:00]     Fetching Certificate (aggregator)...     
DEBUG[2020-03-09T15:41:22-04:00]     Generating Certificate (aggregator)...   
DEBUG[2020-03-09T15:41:23-04:00]     Fetching Certificate (aggregator-ca-bundle)... 
DEBUG[2020-03-09T15:41:23-04:00]       Fetching Certificate (aggregator-signer)... 
DEBUG[2020-03-09T15:41:23-04:00]       Generating Certificate (aggregator-signer)... 
DEBUG[2020-03-09T15:41:23-04:00]     Generating Certificate (aggregator-ca-bundle)... 
DEBUG[2020-03-09T15:41:23-04:00]     Fetching Certificate (system:kube-apiserver-proxy)... 
DEBUG[2020-03-09T15:41:23-04:00]       Fetching Certificate (aggregator-signer)... 
DEBUG[2020-03-09T15:41:23-04:00]       Reusing previously-fetched Certificate (aggregator-signer) 
DEBUG[2020-03-09T15:41:23-04:00]     Generating Certificate (system:kube-apiserver-proxy)... 
DEBUG[2020-03-09T15:41:23-04:00]     Fetching Certificate (aggregator-signer)... 
DEBUG[2020-03-09T15:41:23-04:00]     Reusing previously-fetched Certificate (aggregator-signer) 
DEBUG[2020-03-09T15:41:23-04:00]     Fetching Certificate (system:kube-apiserver-proxy)... 
DEBUG[2020-03-09T15:41:23-04:00]       Fetching Certificate (aggregator)...   
DEBUG[2020-03-09T15:41:23-04:00]       Reusing previously-fetched Certificate (aggregator) 
DEBUG[2020-03-09T15:41:23-04:00]     Generating Certificate (system:kube-apiserver-proxy)... 
DEBUG[2020-03-09T15:41:23-04:00]     Fetching Certificate (etcd-ca-bundle)... 
DEBUG[2020-03-09T15:41:23-04:00]     Reusing previously-fetched Certificate (etcd-ca-bundle) 
DEBUG[2020-03-09T15:41:23-04:00]     Fetching Certificate (etcd-metric-ca-bundle)... 
DEBUG[2020-03-09T15:41:23-04:00]     Reusing previously-fetched Certificate (etcd-metric-ca-bundle) 
DEBUG[2020-03-09T15:41:23-04:00]     Fetching Certificate (etcd-metric-signer)... 
DEBUG[2020-03-09T15:41:23-04:00]     Reusing previously-fetched Certificate (etcd-metric-signer) 
DEBUG[2020-03-09T15:41:23-04:00]     Fetching Certificate (etcd-metric-signer-client)... 
DEBUG[2020-03-09T15:41:23-04:00]     Reusing previously-fetched Certificate (etcd-metric-signer-client) 
DEBUG[2020-03-09T15:41:23-04:00]     Fetching Certificate (etcd-signer)...    
DEBUG[2020-03-09T15:41:23-04:00]     Reusing previously-fetched Certificate (etcd-signer) 
DEBUG[2020-03-09T15:41:23-04:00]     Fetching Certificate (etcd-client)...    
DEBUG[2020-03-09T15:41:23-04:00]     Reusing previously-fetched Certificate (etcd-client) 
DEBUG[2020-03-09T15:41:23-04:00]     Fetching Certificate (journal-gatewayd)... 
DEBUG[2020-03-09T15:41:23-04:00]       Fetching Root CA...                    
DEBUG[2020-03-09T15:41:23-04:00]       Reusing previously-fetched Root CA     
DEBUG[2020-03-09T15:41:23-04:00]     Generating Certificate (journal-gatewayd)... 
DEBUG[2020-03-09T15:41:24-04:00]     Fetching Certificate (kube-apiserver-lb-ca-bundle)... 
DEBUG[2020-03-09T15:41:24-04:00]     Reusing previously-fetched Certificate (kube-apiserver-lb-ca-bundle) 
DEBUG[2020-03-09T15:41:24-04:00]     Fetching Certificate (kube-apiserver-external-lb-server)... 
DEBUG[2020-03-09T15:41:24-04:00]       Fetching Certificate (kube-apiserver-lb-signer)... 
DEBUG[2020-03-09T15:41:24-04:00]       Reusing previously-fetched Certificate (kube-apiserver-lb-signer) 
DEBUG[2020-03-09T15:41:24-04:00]       Fetching Install Config...             
DEBUG[2020-03-09T15:41:24-04:00]       Reusing previously-fetched Install Config 
DEBUG[2020-03-09T15:41:24-04:00]     Generating Certificate (kube-apiserver-external-lb-server)...

patrickdillon commented:
I don't think this is conceptually different than the previous output snippet. You made changes (e.g. [here](https://github.com/openshift/installer/pull/3263/files#diff-266f03122200ace2d21f2e449b6a5503R117)) that would appear in installer output at the standard level of verbosity. I would like to see that output. When you don't specify `--log-level debug` installer output is not very long, so you could just paste all of it.

rna-afk commented:
it's just that it is creating it...need some time for it to finish sorry...I'll post it here when it's done

rna-afk commented:
Removed the part that prints the timer information in debug logs. Changed the timer object to pick and store all the timer information in a package scoped variable and used the stdlib Duration object for the display.

The system now puts up a summary at the end of the installer command specifying all the stages and the time it took for each stage. Currently supports infrastructure, bootstrap, api, operators and bootstrap destroy stages.

Sample logs:
```
INFO Creating infrastructure resources...
INFO Waiting up to 20m0s for the Kubernetes API at https://api.qwerty.gcp.devcluster.openshift.com:6443...
INFO API v1.17.1 up
INFO Waiting up to 40m0s for bootstrapping to complete...
INFO Destroying the bootstrap resources...
INFO Waiting up to 30m0s for the cluster at https://api.qwerty.gcp.devcluster.openshift.com:6443 to initialize...
INFO Waiting up to 10m0s for the openshift-console route to be created...
INFO Install complete!
INFO To access the cluster as the system:admin user when using 'oc', run 'export KUBECONFIG=/home/anarayan/go/src/github.com/openshift/installer/bin/test/auth/kubeconfig'
INFO Access the OpenShift web-console here: https://console-openshift-console.apps.qwerty.gcp.devcluster.openshift.com
INFO Login to the console with user: kubeadmin, password: gannn-LS8Wr-ydpSc-dSDtR
INFO Time elapsed for each section
INFO 	Infrastructure : 2m53s
INFO 	API : 5m45s
INFO 	Bootstrap Complete : 12m48s
INFO 	Bootstrap Destroy : 3m16s
INFO 	Cluster operator : 3m44s
INFO Total time elapsed : 28m26s
```

rna-afk commented:
> What's the goal of the change?
The goal of the change is that now the users can see exactly which stage took how much time to complete so they can plan the installation at a later stage.
> If it's tracking stages like infrastructure creation , bootstrap wait, bootstrap destroy, cluster operator install complete. I don't see how this addresses that completely?
Changed the logic. Hopefully, it will now.
> The timer is conflating all assets together.
Changed the logic.
> We are sending another metrics timer and it still requires manual start stop elapsed. What is that package providing value?
Right now after the change, the timer is a one-stop shop where all the information about each stage is captured and logged. This can be reused by other systems like the metrics package to get all the information that is needed.

patrickdillon commented on pkg/metrics/timer.go:
```diff
@@ -0,0 +1,42 @@
+	startTime    time.Time
+	totalTime    time.Time
+	listOfStages []string
+	listOfTimes  []time.Duration
```
We shouldn't rely on the index of listOfTimes to correspond to listOfStages. For one, this will break down if we want to use this in a different command, such as wait-for or destroy.

rna-afk replied:
Was worried about the order in which the error messages were printed and hence used a list.

patrickdillon commented on pkg/metrics/timer.go:
```diff
@@ -0,0 +1,42 @@
+	for index, item := range timer.listOfStages {
+		logrus.Infof(indent+"%s : %v", item, timer.listOfTimes[index])
+	}
+	logrus.Infof("Total time elapsed : %v", time.Now().Round(time.Second).Sub(timer.totalTime))
```
We're not storing total duration. How would this integrate with pushing metrics to the telemeter?

rna-afk replied:
The code here is only for the logging. This function can be the place where the duration metrics is populated. Storing the start time seemed enough as the total time can be deduced rather than stored.

patrickdillon replied:
This PR should be created with future metrics collection in mind. We should try to create something that can either be extended or used by a client to achieve the metrics use case. If we  knowingly need to change this for metrics, we should try to address that now.

patrickdillon commented on pkg/metrics/timer.go:
```diff
@@ -0,0 +1,42 @@
+	timer = Timer{}
+	timer.startTime = time.Now().Round(time.Second)
+	timer.totalTime = time.Now().Round(time.Second)
+	timer.listOfStages = []string{"Infrastructure", "API", "Bootstrap Complete", "Bootstrap Destroy", "Cluster operator"}
```
operator -> Operator. Maybe we could think of a better name for this. Operators Available  or Cluster Operators Available perhaps.

rna-afk replied:
Wasn't sure exactly what to print here so I'm all ears.

patrickdillon commented on pkg/metrics/timer.go:
```diff
@@ -0,0 +1,60 @@
+	logrus.Debugf("Time elapsed for each section")
+	indent := "	"
+	for _, item := range timer.listOfStages {
+		logrus.Debugf(indent+"%s	:	%v", item, timer.stageTimes[item])
```
I think there may be some tabs in here which are showing up when printed in CI. See: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/3263/pull-ci-openshift-installer-master-e2e-aws/10586/artifacts/e2e-aws/installer/.openshift_install.log

Use spaces instead of tabs.

patrickdillon commented on pkg/metrics/timer.go:
```diff
@@ -0,0 +1,60 @@
+	logrus.Debugf("Time elapsed for each section")
+	indent := "	"
+	for _, item := range timer.listOfStages {
+		logrus.Debugf(indent+"%s	:	%v", item, timer.stageTimes[item])
```
There should not be a space between the left-hand side and the colon. Also I think the duration values should be left justified.

rna-afk replied:
Fixing the spaces properly this time.

patrickdillon commented on pkg/metrics/timer.go:
```diff
@@ -0,0 +1,60 @@
+
+// LogSummary prints the summary of all the times collected so far into the INFO section.
+func LogSummary() {
+	logrus.Debugf("Time elapsed for each section")
```
This should not be printed when destroy or any other command with no stages runs.

rna-afk replied:
Well this function will only have to be called during the exit of the command that we want to log.

patrickdillon replied:
Currently, if you run `openshift-install destroy cluster` the output is:

```
Time elapsed for each section
Total time elapsed : 4m29s"
```
"Time elapsed for each section" is confusing here and should not be printed when destory or another command with no stages is run.

There should be a colon at the end. Also consider playing around with the wording, perhaps just `Time elapsed:` It is probably better to avoid the word "section" in an official capacity.

rna-afk commented:
> I left a few comments. I think it would be really good to have a sense of how metrics collection would tie into this. 

The tie-up will be to pull the values from the struct in the LogSummary function and put it into the labels map of the duration. The call to push will be made afterwards if needed. 

> Also, the PR description is outdated as some of the previous functionality was removed.

Fixed it

patrickdillon commented:
>The tie-up will be to pull the values from the struct in the LogSummary function and put it into the labels map of the duration. The call to push will be made afterwards if needed.

I would like to see further thought on this not only for the reasons I [listed above](https://github.com/openshift/installer/pull/3263#discussion_r393040911), but also:

- on its face it would seem better to separate these responsibilities. A function that prints is doing enough already without adding another responsibility of adding labels
- Let's say we do extend this function to add labels. How would that work? e.g. how do we know which metric to add the value to? Let's say we wanted to add the total time for the destroy command, how can we use the information in this function to do that? The only way I see that we can draw a connection between the destroy command and the metrics here is to look for an empty map.

rna-afk commented:
Aligned the display. Display is now like this.

time="2020-03-16T13:44:52-04:00" level=debug msg="Time elapsed for each section"
time="2020-03-16T13:44:52-04:00" level=debug msg="Infrastructure             :  3m4s"
time="2020-03-16T13:44:52-04:00" level=debug msg="API                            :  5m2s"
time="2020-03-16T13:44:52-04:00" level=debug msg="Bootstrap Complete  :  11m6s"
time="2020-03-16T13:44:52-04:00" level=debug msg="Bootstrap Destroy     :  3m11s"
time="2020-03-16T13:44:52-04:00" level=debug msg="Cluster Operator       :  6m41s"
time="2020-03-16T13:44:52-04:00" level=debug msg="Console                    :  0s"
time="2020-03-16T13:44:52-04:00" level=info msg="Time elapsed : 29m4s"
```

rna-afk commented:
> Let's say we do extend this function to add labels. How would that work? e.g. how do we know which metric to add the value to? Let's say we wanted to add the total time for the destroy command, how can we use the information in this function to do that? The only way I see that we can draw a connection between the destroy command and the metrics here is to look for an empty map.

Right. The other way to think about it would be to assume the timer has the values and the metrics must pick the values from this timer and populate its labels based on that. Could be one of the functionality of the builder's build() function to take the timer in and pick the values up from it.

rna-afk commented on pkg/metrics/timer.go:
```diff
@@ -0,0 +1,71 @@
+package metrics
```
I'm a little unclear about sub packages. In Go would the metrics package still have access to it? If it does not then there must be an extra function that gives access to the data collected.

patrickdillon replied:
> would the metrics package still have access to it?

Only if the metrics package imports it and the fields are exported, which as you mentioned, is not the way to go.

>  If it does not then there must be an extra function that gives access to the data collected

Instead, timer would import the metrics package to get the `SetValue` function and the metrics name constants (those could be moved from timer to metrics later). Metrics package would not have any dependency on the timer package.

rna-afk replied:
> Instead, timer would import the metrics package to get the SetValue function and the metrics name constants (those could be moved from timer to metrics later). Metrics package would not have any dependency on the timer package.

yeah this makes sense. It was either the timer using metrics or vice versa. I'll change this then.

rna-afk commented:
Changed the package name and will add the SetValue function along with the metrics package

patrickdillon commented on pkg/metrics/timer/timer.go:
```diff
@@ -0,0 +1,71 @@
+	}
+
+	for _, item := range timer.listOfStages {
+		logrus.Debugf(indent+"%-"+strconv.Itoa(maxLen)+"s  :  %v", item, timer.stageTimes[item])
```
I think the colon should be "attached" to the left hand word, not in its own column, but that might just be my taste.

rna-afk replied:
I'm not sure? but I usually leave a space between everything possible so I feel it looks neat. I think it's a preference? gofmt did not complain.

rna-afk commented:
>That sounds good. Any plans to address:
I think this wold still need some adjustments to set the total time for the appropriate metric (create vs wait-for vs destroy).

Was thinking the same approach as SetTimeElapsed and SetValue except use the Push/PushAll function to pick the time and set it before build and push.

patrickdillon commented:
> Was thinking the same approach as SetTimeElapsed and SetValue except use the Push/PushAll function to pick the time and set it before build and push.

Why treat total time differently than other times? If possible, I think it's better to treat them all uniformly,

> use the Push/PushAll function to pick the time and set it before build and push.

I think we should try to follow a pattern where all metric and label values are registered using the metrics pkg api, so in this case `SetValue(MetricName, Value)` or something to that effect. This pkg should act as a client of the metrics package api, metrics package should not depend on this duration package.

rna-afk commented:
> Why treat total time differently than other times? If possible, I think it's better to treat them all uniformly,

I was trying to reduce calls to the pkg from the installer and thought the LogSummary function can record the time. The idea was to be as less invasive in the installer code as possible.

> I think we should try to follow a pattern where all metric and label values are registered using the metrics pkg api, so in this case `SetValue(MetricName, Value)` or something to that effect. This pkg should act as a client of the metrics package api, metrics package should not depend on this duration package.

Hmm...was thinking about keeping the section that prints the time and the metrics section separate. If we combine them, then your suggestion makes sense and the way the times are collected has to be changed then.

patrickdillon commented:
> I was trying to reduce calls to the pkg from the installer and thought the LogSummary function can record the time. The idea was to be as less invasive in the installer code as possible.

Yes this is a good goal and you've done a very good job of that so far. There are many ways you can maintain this goal without creating the [problem I am pointing out](https://github.com/openshift/installer/pull/3263#pullrequestreview-379809570). 

> Right now, my thought was to call LogSummary when SetValue is called but I shouldn't do that. I'll add the total time collection section instead of getting it in the LogSummary.

Yeah I think it should be the reverse, LogSummary calls SetValue. Or LogSummary calls SetTimeElapsed which calls SetValue. This second suggestion would use an insignificant amount of extra memory and would keep it more DRY. Both cases would need the metric name passed to LogSummary.

rna-afk commented:
Added sections to the installer to add the total time and not to automatically generate the total time in the LogSummary function to allow the metrics system to use this module with more flexibility.

abhinavdahiya commented on pkg/metrics/timer/timer.go:
```diff
@@ -0,0 +1,76 @@
+}
+
+const (
+	// InfrastructureStageName is a constant string value that is used to display the Infrastructure created stage text in the logs.
+	InfrastructureStageName = "Infrastructure"
+	// APIStageName is a constant string value that is used to display the API stage text in the logs.
+	APIStageName = "API"
+	// BootstrapCompleteStageName is a constant string value that is used to display the Bootstrap Complete text in the logs.
+	BootstrapCompleteStageName = "Bootstrap Complete"
+	// BootstrapDestroyStageName is a constant string value that is used to display the Boostrap Destroy Complete text in the logs.
+	BootstrapDestroyStageName = "Bootstrap Destroy"
+	// ClusterOperatorStageName is a constant string value that is used to display the Cluster Operator Complete text in the logs.
+	ClusterOperatorStageName = "Cluster Operator"
+
+	// ConsoleStageName is a constant string value that is used to display the Cluster Operator Complete text in the logs.
+	ConsoleStageName = "Console"
```
why does the timer package get to control what stages can exist?

rna-afk replied:
These are just constants to use. I've kept them here as they were relevant to the timer.

abhinavdahiya replied:
I want this package to not provide these const. I think this is un necessarily restricts the users with what stages they can use.

abhinavdahiya replied:
I think the only one I think that can should stay here is Total

abhinavdahiya commented on pkg/metrics/timer/timer.go:
```diff
@@ -0,0 +1,76 @@
+package timer
```
so personally, i feel the goal of this package should be `allow installer codebase to track various stages/units of work` 
another useful api provided would be a `Summarize` function that can summarize all the stages.

Now to achieve that, we should allow the user (here is could an asset, or the cmd package) to start a specific stage and end a specific stage easily, but still allowing another user (it could be the same asset, or cmd package)
so whoever holds the pointer to the timer-tracker can summarize all the timers.

so I would prefer this package api to look like

```
// global tracker to reduce the _how to distribute the timer_ problem for now
StartTimer(key string)
StopTimer(key string)
Summarize(something to control the output location)

// also allow some people to create an explicit ones for other use.
NewTracker() Tracker
Tracker struct
(Tracker) StartTimer(key string)
(Tracker) StopTimer(key string)
(Tracker) Summarize(...)

```

patrickdillon replied:
@abhinavdahiya this design sounds good to me. one of the things that @rna-afk and I have been discussing is how these timer values will be attached to the metrics created in the metrics package. 

Do you have an opinion on whether:

1. The "user" pushes the value to metrics package after reading from this tracker
2. The tracker itself pushes the value (based on key)
3. Metrics package pulls values from this API

Looking at the suggested API I think the first option, which sounds good to me. I have been trying to avoid option 3.

abhinavdahiya replied:
> 1. The "user" pushes the value to metrics package after reading from this tracker

I like this option, it will allows us to write the glue code that will extract these from tracker and push to metrics. so both these are separate and don't depend on one another.

patrickdillon replied:
was discussing with @rna-afk and I'm not sure I get why we need a global tracker at all? 

> // global tracker to reduce the _how to distribute the timer_ problem for now

I'm trying to think of a reason in case we eventually we made certain sections concurrent, but I'm not sure that use case would make sense either.

abhinavdahiya replied:
> // also allow some people to create an explicit ones for other use.
NewTracker() Tracker
Tracker struct
(Tracker) StartTimer(key string)
(Tracker) StopTimer(key string)
(Tracker) Summarize(...)

in my original comment, tracker tracks multiple timers. so if you have access to a locally defined tracker for example, an asset creating one for tracking various stages internally, the asset can start stop timers.

but for us today, esp in cmd/ there isn't a great place to define a tracker that can use used by everyone to start and stop their timers. 

> // global tracker to reduce the _how to distribute the timer_ problem for now
StartTimer(key string)
StopTimer(key string)
Summarize(something to control the output location)

That's where a global tracker comes into play where we can import the package and use the global tracker to interact.

that was my thought. hopefully that provides more context. @patrickdillon @rna-afk

rna-afk replied:
With regards to the input parameter of the Summarize function, am I right to think that the input would be something in the lines of debug levels or is it something more?

rna-afk commented:
Made the changes to the timer based on Abhinav's comments. Extended the Timer type to start and stop specific stages and added a NewTimer function that returns a Timer object that can be used for specific tracking. A global Timer object is maintained to keep track of all the sections during installation right now.

patrickdillon commented on pkg/metrics/timer/timer.go:
```diff
@@ -0,0 +1,123 @@
+}
+
+// StartTimer initiailzes the timer object with the current timestamp information.
+func StartTimer(key string) {
```
Can we reuse the functions with the `Timer` pointer receivers so we don't have two copies of all the functions? Can we use `NewTimer` to make the global timer?

rna-afk replied:
> Can we reuse the functions with the Timer pointer receivers so we don't have two copies of all the functions? Can we use NewTimer to make the global timer?

Actually I have changed these and committed but not pushed yet. Waiting on Abhinav's comment on what the input should be for the Summarize function should look like and I'll push them together.

abhinavdahiya commented on pkg/metrics/timer/timer.go:
```diff
@@ -0,0 +1,97 @@
+}
+
+// LogSummary prints the summary of all the times collected so far into the INFO section.
+func (t *Timer) LogSummary() {
```
I would like if you can add a unit test for this.

And to safely unit test, this needs to get explicit input of where to log and not use the global logger here in this func or the Timer struct itself.

abhinavdahiya replied:
Secondly can you add comments as how this function generates the summary?

patrickdillon commented on pkg/metrics/timer/timer.go:
```diff
@@ -0,0 +1,97 @@
+
+	for _, item := range t.listOfStages {
+		if item != TotalTimeElapsed {
+			logrus.Debugf(indent+"%-"+strconv.Itoa(maxLen)+"s  :  %v", item, t.stageTimes[item])
```
I think this line could be cleaned up to make it more readable:

- `indent` appears to prepend an empty string, so can we just drop it?
- use a format string like `Sprintf` (in general string concatenation should be avoided)
- using `*` you can pass in a width as an arg to `Sprintf`, see [here](https://stackoverflow.com/questions/42308070/how-to-dynamically-pass-values-to-sprintf-or-printf) and [here](https://golang.org/pkg/fmt/)
- I still think `:` having its own column is not good

rna-afk commented:
- Added comments to the LogSummary function and also added the logrus.Logger interface to the LogSummary parameters to control the location of output for the function.
- Changed the formatting of the LogSummary with regards to the output.
- Added basic unit test cases.

abhinavdahiya commented on cmd/openshift-install/create.go:
```diff
@@ -352,6 +364,7 @@ func waitForInitializedCluster(ctx context.Context, config *rest.Config) error {
 	defer cancel()
 
 	failing := configv1.ClusterStatusConditionType("Failing")
+	timer.StartTimer("Cluster Operator")
```
`Cluster Operator` -> `Cluster Operators`

rna-afk replied:
Fixed it.

abhinavdahiya commented on pkg/metrics/timer/timer_test.go:
```diff
@@ -0,0 +1,130 @@
+	logger := logrus.New()
+	logger.Out = &textOutput
+	logger.Level = logrus.DebugLevel
+	logger.Formatter = &logrus.JSONFormatter{}
```
why the JSONFormatter?

rna-afk replied:
With JSONFormatter, it was easier to pick just the message that the logger was producing rather than the TextFormatter where I had to slice the strings to pull it. Could have created a struct and parsed it into the struct but thought JSON was the same as creating a struct so thought of using that.

abhinavdahiya commented on pkg/metrics/timer/timer.go:
```diff
@@ -0,0 +1,95 @@
+
+	for _, item := range t.listOfStages {
+		if item != TotalTimeElapsed {
+			logger.Debugf(fmt.Sprintf("%*s: %v", maxLen, item, t.stageTimes[item]))
```
`%v` ?? why is it not `%s`

rna-afk replied:
%v was designed specifically for formatting Go types but in this case it does exactly the same as %s. Changed it.

abhinavdahiya commented on pkg/metrics/timer/timer.go:
```diff
@@ -0,0 +1,95 @@
+// StopTimer records the duration for the current stage sent as the key parameter and stores the information.
+func (t *Timer) StopTimer(key string) {
+	if item, found := t.startTimes[key]; found {
+		duration := time.Now().Round(time.Second).Sub(item).Round(time.Second)
```
https://golang.org/pkg/time/#Since

rna-afk replied:
Seems better. Changed the code to use it.

abhinavdahiya commented on pkg/metrics/timer/timer_test.go:
```diff
@@ -0,0 +1,130 @@
+	timer.StartTimer("testStage3")
+	timer.StartTimer("testStage4")
+
+	// time.Sleep(5 * time.Second)
```
I don't think we should keep commented code.

rna-afk replied:
Fixed it.

abhinavdahiya commented:
/retest

abhinavdahiya commented:
/test e2e-metal

abhinavdahiya commented:
/test e2e-gcp-upi

abhinavdahiya commented:
/test e2e-azure

abhinavdahiya commented:
/test e2e-gcp

abhinavdahiya commented:
/test e2e-metal-ipi

abhinavdahiya commented:
/approve

patrickdillon commented:
I would like to see some indication of how this would be used with metrics package according to https://github.com/openshift/installer/pull/3263#discussion_r399554408 

How will the caller get the duration value to push to metrics? `StopTimer` could potentially return a value along with many other options

rna-afk commented:
> I would like to see some indication of how this would be used with metrics package according to [#3263 (comment)](https://github.com/openshift/installer/pull/3263#discussion_r399554408)
> 
> How will the caller get the duration value to push to metrics? `StopTimer` could potentially return a value along with many other options

I was thinking of the two approaches, one of them is calling a new function in the metrics system ("UpdateDurationMetric(stageName string, duration Duration)") to update the corresponding metric with the duration value. This update will only occur when the global StopTimer function is called. 

I was thinking of keeping the timer package and the metrics system separate as the timer has the extra functionality of logging it into the console. If you think they belong together, then the signature of the global StopTimer will have to change to include an extra parameter, metricName and that in turn will call the UpdateDurationMetric function with the metricName and the duration and update the corresponding duration metric.

Was hoping to experiment on this in my local branch after this PR gets merged. Did not want to include the metrics changes with this PR.

patrickdillon commented:
> Was hoping to experiment on this in my local branch after this PR gets merged. Did not want to include the metrics changes with this PR.

We intend to use this package in combination with the metrics package, so I think a timer API that is compatible with intended metrics design is a requirement. 

> one of them is calling a new function in the metrics system

> the signature of the global StopTimer will have to change to include an extra parameter, metricName

Neither of these suggestions are in-line with my [original comment](https://github.com/openshift/installer/pull/3263#issuecomment-608423119) or the link therein. Can you take another look at the original comment and link and let me know if the suggestions are unclear?

rna-afk commented:
> Neither of these suggestions are in-line with my [original comment](https://github.com/openshift/installer/pull/3263#issuecomment-608423119) or the link therein. Can you take another look at the original comment and link and let me know if the suggestions are unclear?

I think I get it now. The StopTimer returns the duration value which the installer picks up and uses it to update the metrics. I'll change the StopTimer function to return a Duration object which can be used.

patrickdillon commented on cmd/openshift-install/create.go:
```diff
@@ -418,6 +432,7 @@ func waitForConsole(ctx context.Context, config *rest.Config, directory string)
 	// no route in a row (to show we're still alive).
 	logDownsample := 15
 	silenceRemaining := logDownsample
+	timer.StartTimer("Console")
```
this is (accurately) showing `0s` in the fips CI run. not sure if there are other cases where console may take longer to come up.

rna-afk replied:
I have never seen a value other than 0s during my tests. Not sure if it's a concern.

patrickdillon commented:
/test e2e-aws-upi
/test e2e-gcp-upi
/test e2e-azure-upi
/test e2e-vsphere-upi

patrickdillon commented on pkg/metrics/timer/timer.go:
```diff
@@ -0,0 +1,96 @@
+		}
+	}
+
+	if maxLen != 0 {
```
If we ditch `API` or `Console` I think this check may end up with output like this:

```
Time elapsed for each section
Cluster Operators: 9m13s
Time elapsed: 9m13s
```
It might be better to just say `Time elapsed: xxx` if there is only one section, but this is not a blocker for me.

patrickdillon commented on pkg/metrics/timer/timer.go:
```diff
@@ -0,0 +1,96 @@
+	}
+
+	if maxLen != 0 {
+		logger.Debugf("Time elapsed for each section")
```
I feel like "section" is not great. How about "Time elapsed for each section" -> "Time elapsed per stage:"?

patrickdillon commented:
After discussion in slack we decided to leave in API and console. I will lgtm after checking latest changes in ci
/test e2e-aws-upi
/test e2e-gcp-upi
/test e2e-azure-upi
/test e2e-vsphere-upi

@rna-afk I am not sure if this PR satisfies either or both of the github issues in the jira card. Can you check them out and if it does satisfy one or both can you link them to this PR so that they close when this PR merges?

rna-afk commented:
resolves #2942 
resolves #2943

patrickdillon commented:
/lgtm
